crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,432 @@
+use crate::ports::P2pDb;
+use dashmap::DashMap;
+use fuel_core_metrics::p2p_metrics::{
+    increment_p2p_req_res_cache_hits,
+    increment_p2p_req_res_cache_misses,
+};
+use fuel_core_storage::Result as StorageResult;
+use fuel_core_types::{
+    blockchain::SealedBlockHeader,
+    services::p2p::Transactions,
+};
+use std::{
+    collections::VecDeque,
+    hash::Hash,
+    ops::Range,
+    sync::{
+        Arc,
+        Mutex,
+    },
+};
+
+type BlockHeight = u32;
+
+struct LruCache<K, V> {
+    cache: DashMap<K, Arc<V>>,
+    order: Mutex<VecDeque<K>>,
+    capacity: usize,
+}
+
+impl<K, V> LruCache<K, V>
+where
+    K: Eq + Hash + Clone,
+{
+    fn new(capacity: usize) -> Self {
+        Self {
+            cache: DashMap::new(),
+            order: Mutex::new(VecDeque::new()),
+            capacity,
+        }
+    }
+
+    fn insert(&self, key: K, value: V) {
+        let mut order = self.order.lock().expect("Poisoned lock");
+
+        if self.cache.len() >= self.capacity {
+            if let Some(least_used) = order.pop_front() {
+                self.cache.remove(&least_used);
+            }
+        }
+
+        self.cache.insert(key.clone(), Arc::new(value));
+
+        // Update the access order.
+        order.retain(|k| k != &key);
```

Comment:
I meant to use already existing implementation of the LRU from crates.io. The current implementation is too slow. and operation here is expensive. Plus, usage of the `Mutex` to manage `order` destroyed all benefits of using `DashMap`.

Reply:
without using `Dashmap`, we will need a mutable reference to the underlying LruCache, from https://crates.io/crates/lru for example. and for every p2p request that comes through, we will need to pass a mutable reference of the `CachedView` https://github.com/FuelLabs/fuel-core/blob/496071f27230210e9ef448245faf22e0f6b2c885/crates/services/p2p/src/service.rs#L581, which isn't possible without a mutex (which is what I assume you want to avoid). This is why I implemented the lru cache using a dashmap.

Reply:
we could use something like https://docs.rs/crossbeam-skiplist/latest/crossbeam_skiplist/ to order the insertions of elements

Reply:
I don't want lock-free LRU. The `Dashmap` has locks inside, and I'm trying to say that usage of the `Mutex` on upper level removes benefits of using `Dashmap` that uses RW lock inside.

We don't need to re-invent LRU implementation, we just can reuse already existing implementation that works with `&self` without requiring usage of the mutex/rwlock by us explicitly.

Reply:
replaced dashmap with quick_cache which supports LRU-style eviction - 1bdcf0094f